modules/core/src/main/scala/com.snowplowanalytics/iglu.schemaddl/parquet/Migrations.scala

```diff
@@ -0,0 +1,223 @@
+      * `KeyRemoval` historical data loss, such as removing field in Struct. Loading will succeed with data loss.
+      * `IncompatibleType` Incompatible type casting, such as String -> Int, etc. Loading will fail.
+      * `RequiredNullable` Nullable to required field or array migration. Loading will fail.
+
+   Note that JSON type and String are interchangeable, in the context of this library because Spark does not support
+   Parquet with JSON type, so they are stringified.
+ */
+object Migrations {
+
+  // Path to field or type within the Parquet schema structure. String `"[arrayDown]"` used to indicate array traversal
+  type ParquetSchemaPath = List[String]
+
+
+  // Base trait for schema differences, which would result in migration.
+  sealed trait ParquetMigration {
+    // As Migration logic traverses deeper into the structure keys appended to the head of list.
+    // So path stored is reverse.
+    def path: ParquetSchemaPath
+
+    def reversedPath: ParquetSchemaPath = path.reverse
+  }
+
+  type ParquetSchemaMigrations = Set[ParquetMigration]
+
+  sealed trait Patch extends ParquetMigration
+
+  sealed trait Minor extends ParquetMigration
+
+  sealed trait Major extends ParquetMigration
+
+  case class KeyRemoval(override val path: ParquetSchemaPath, removedKey: Type) extends Major {
+    override def toString: String = s"Key removal at /${reversedPath.mkString("/")}"
+  }
+
+  case class RequiredNullable(override val path: ParquetSchemaPath) extends Major {
+    override def toString: String = s"Changing nullable property to required at /${reversedPath.mkString("/")}"
+  }
+
+  case class NullableRequired(override val path: ParquetSchemaPath) extends Patch {
+    override def toString: String = s"Changing required property to nullable at /${reversedPath.mkString("/")}"
+  }
+
+  case class TopLevelKeyAddition(override val path: ParquetSchemaPath, key: Type) extends Minor {
+    override def toString: String = s"Top-level schema key addition at /${reversedPath.mkString("/")}"
+  }
+
+  case class NestedKeyAddition(override val path: ParquetSchemaPath, key: Type) extends Patch {
+    override def toString: String = s"Nested object key addition at /${reversedPath.mkString("/")}"
+  }
+
+  case class TypeWidening(override val path: ParquetSchemaPath, oldType: Type, newType: Type) extends Patch {
+    override def toString: String = s"Type widening from $oldType to $newType at /${reversedPath.mkString("/")}"
+  }
+
+  case class IncompatibleType(override val path: ParquetSchemaPath, oldType: Type, newType: Type) extends Major {
+    override def toString: String = s"Incompatible type change $oldType to $newType at /${reversedPath.mkString("/")}"
+  }
+
+  private implicit class FocusStruct(val value: Struct) {
+    def focus(subkey: String): Option[Field] = value.fields.collectFirst { case x if x.name == subkey => x }
+  }
+
+  private case class MigrationTypePair(path: ParquetSchemaPath, sourceType: Type, targetType: Type) {
```

Comment:
From pov of spark/databricks, it does not matter which schema is the source or which schema is the target.  I think the rules are symmetrical.  E.g. both of these are valid:

1. Current databricks schema has field xyz with type=Integer.  We try to load a parquet file where type=Long.  The databricks schema is updated to type=Long.
2. Current databricks schema has field xyz with type=Long.  We try to load a parquet file where type=Integer.  It just works.

Reply:
Second would produce error on copy the `Failed to merge incompatible data types LongType and IntegerType`. I have tested this.

Reply:
Oh this is bad news!  Imagine schema `1-0-0` has field `xyz: type=int` and schema `1-0-1` has `type=long`.  And this series of events:

- Batch 1 contains `1-0-0` events only.  The databricks table has type=int
- Batch 2 contains a `1-0-1` event, so the table updates to type=long
- Batch 3 contains `1-0-0` events only.

Does this mean Batch 3 will break loading?  Because the table has type=long and the parquet file has type=int.